topiary/src/atom_collection.rs

```diff
@@ -773,63 +773,61 @@ fn detect_multi_line_nodes(node: &Node) -> HashSet<usize> {
 }
 
 fn detect_blank_lines_before(node: &Node) -> HashSet<usize> {
-    detect_line_breaks_inner(node, 2, None, None).0
+    detect_line_breaks_inner(node, 2).0
 }
 
 fn detect_line_break_before_and_after(node: &Node) -> (HashSet<usize>, HashSet<usize>) {
-    let result = detect_line_breaks_inner(node, 1, None, None);
-    (result.0, result.1)
+    detect_line_breaks_inner(node, 1)
 }
 
-// TODO: This is taking a bit too much time, and would benefit from an
-// optimization.
-// TODO 2: The whole function is a mess now, and should be rewritten.
-fn detect_line_breaks_inner(
+pub fn detect_line_breaks_inner(
     node: &Node,
     minimum_line_breaks: u32,
-
-    // TODO: Replace these with just previous_node: Option<&Node>
-    previous_node_id: Option<usize>,
-    previous_end: Option<u32>,
-) -> (HashSet<usize>, HashSet<usize>, Option<usize>, Option<u32>) {
-    let mut nodes_with_breaks_before = HashSet::new();
-    let mut nodes_with_breaks_after = HashSet::new();
-
-    if let (Some(previous_node_id), Some(previous_end)) = (previous_node_id, previous_end) {
-        let current_start = node.start_position().row();
-
-        if current_start >= previous_end + minimum_line_breaks {
-            nodes_with_breaks_before.insert(node.id());
-            nodes_with_breaks_after.insert(previous_node_id);
-
-            log::debug!(
-                "There are at least {} blank lines between {:?} and {:?}",
-                minimum_line_breaks,
-                previous_node_id,
-                node.id()
-            );
+) -> (HashSet<usize>, HashSet<usize>) {
+    // Flatten the tree, depth-first, into a vector of nodes
+    let mut walker = node.walk();
+    let mut dfs_nodes: Vec<Node> = Vec::new();
+
+    // NOTE Could this be written in functional style?
+    // Either way, it can be factored out to avoid running twice.
+    'walk: loop {
```

Comment:
It looks cleaner to do this recursively, like we do several places, e.g. `collect_leafs_inner`.

Reply:
I am factoring this out in a subsequent PR. I'll look into making it recursive there...